# LetsGrowMore - Data Science intern

## Task 1 - Iris Flower Classification ML Project

## Author - Vishal Dnyaneshwar Mundhe

## Loading the data set

In [125]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [126]:
df=sns.load_dataset("iris")
df

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

In [127]:
df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [128]:
df.tail()

sepal_length  sepal_width  petal_length  petal_width    species
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

In [129]:
df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

## Checking for missing values

In [130]:
df.isnull().sum()

sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64

## Label Encoding

In [131]:
from sklearn import preprocessing
le=preprocessing.LabelEncoder()
le.fit(df["species"])

LabelEncoder()

## Data separation

In [132]:
x=df.iloc[:,0:4]
x

sepal_length  sepal_width  petal_length  petal_width
0             5.1          3.5           1.4          0.2
1             4.9          3.0           1.4          0.2
2             4.7          3.2           1.3          0.2
3             4.6          3.1           1.5          0.2
4             5.0          3.6           1.4          0.2
..            ...          ...           ...          ...
145           6.7          3.0           5.2          2.3
146           6.3          2.5           5.0          1.9
147           6.5          3.0           5.2          2.0
148           6.2          3.4           5.4          2.3
149           5.9          3.0           5.1          1.8

[150 rows x 4 columns]

In [133]:
y=df.iloc[:,-1]
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: species, Length: 150, dtype: object

## Train test Split

In [134]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2)
x_train

sepal_length  sepal_width  petal_length  petal_width
74            6.4          2.9           4.3          1.3
2             4.7          3.2           1.3          0.2
136           6.3          3.4           5.6          2.4
111           6.4          2.7           5.3          1.9
102           7.1          3.0           5.9          2.1
..            ...          ...           ...          ...
45            4.8          3.0           1.4          0.3
123           6.3          2.7           4.9          1.8
26            5.0          3.4           1.6          0.4
61            5.9          3.0           4.2          1.5
4             5.0          3.6           1.4          0.2

[120 rows x 4 columns]

## KNN Classifier

In [135]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
knn=KNeighborsClassifier(n_neighbors=10)
knn.fit(x_train, y_train)
y_pred=knn.predict(x_test)

C:\vm\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


## Accuracy

In [136]:
acc1=accuracy_score(y_test,y_pred)
acc1

0.9666666666666667

## Confusion Matrix

In [137]:
from sklearn.metrics import confusion_matrix
c_m=confusion_matrix(y_test,y_pred)
c_m

array([[11,  0,  0],
       [ 0, 10,  1],
       [ 0,  0,  8]], dtype=int64)

## Classification Report

In [138]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        11
  versicolor       1.00      0.91      0.95        11
   virginica       0.89      1.00      0.94         8

    accuracy                           0.97        30
   macro avg       0.96      0.97      0.96        30
weighted avg       0.97      0.97      0.97        30



## Random Forest Classifier

In [139]:
from sklearn.ensemble import RandomForestClassifier
clf=RandomForestClassifier(n_estimators=10)
clf.fit(x_train,y_train)
y_pred=clf.predict(x_test)

## Accuracy

In [140]:
acc2=accuracy_score(y_test,y_pred)
acc2

0.9

##  Confusion Matrix

In [141]:
cm=confusion_matrix(y_test,y_pred)
cm

array([[11,  0,  0],
       [ 0,  8,  3],
       [ 0,  0,  8]], dtype=int64)

## Classification Report 

In [143]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        11
  versicolor       1.00      0.73      0.84        11
   virginica       0.73      1.00      0.84         8

    accuracy                           0.90        30
   macro avg       0.91      0.91      0.89        30
weighted avg       0.93      0.90      0.90        30



## Conclusion :

#### I have performed Iris Flower Classification using two classifiers namely KNN ans Random Forest Classifier. Accuracy of KNN Classifiers is 96% and Random Forest Classifier is 90%. Hence we concluded that KNN Classifieris better for classification of Iris flower.